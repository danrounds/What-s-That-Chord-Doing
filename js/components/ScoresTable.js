import React from 'react';
import {connect} from 'react-redux';

import * as actions from '../actions';

export class ScoresTable extends React.Component {

    render() {
        if (this.props.api.pending)
            return <div>'Loading...'</div>;
        else if (!this.props.api.highScores.length)
            return <div>'No relevant scores posted'</div>;
        else if (this.props.api.error)
            return <div>'Error loading scores'</div>;
            
        let entries = [];
        const scoresArray = this.props.api.highScores;

        const mode = this.props.api.lastHighScoreAccessed
                  || this.state.mode + (this.state.inversions ? 'Inv' : '');

        let key = 0;
        for (let i in scoresArray) {
            let name = scoresArray[i].name;
            let { nAnsweredRight, nQuestionNumber, totalClicks, winRatio }
                    = scoresArray[i].scores[mode];

            winRatio = String(winRatio * 100 || '').split('.');
            if (winRatio.length === 2)
                winRatio = [winRatio[0], '.', winRatio[1].substr(0,2)].join('');

            entries.push(
                <tr key={key++}>{[name,nAnsweredRight,nQuestionNumber,totalClicks,winRatio]
                  .map((el) => <th key={key++}>{el}</th>)}
                </tr>
            );
        }
        return (

            <table style={{width: '100%'}}>
              <thead>
                <tr>{['Name','# Right','# Total','Guesses:Right','% Right']
                  .map((el) => <th key={el}>{el}</th>)}
                </tr>
              </thead>
              <tbody>{entries}</tbody>
            </table>

        );
    }
}

const mapStateToProps = (state, props) => ({
    // gameType: state.api.lastHighScoreAccessed || state.game.gameType || 'allChords',
    // inv: state.game.inversions,
    api: state.api || { highScores: [] },
});

export default connect(mapStateToProps)(ScoresTable);
