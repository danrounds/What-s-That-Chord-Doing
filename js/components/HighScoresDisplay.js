import React from 'react';
import {connect} from 'react-redux';

import * as actions from '../actions';
import ScoresTable from './ScoresTable';

export class HighScoresDisplay extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
        this.clickBigButton = this.clickBigButton.bind(this);
        this.bigButtonIsSeleted = this.bigButtonIsSelected.bind(this);
        this.isTriggered = this.isTriggered.bind(this);
        this.trigger = this.trigger.bind(this);
    }

    bigButtonIsSelected(selected) {
        if (selected === this.state.showButtonSubset)
            return { backgroundColor: 'orange' };
        return {};
    }

    clickBigButton(buttonType) {
        this.setState({ showButtonSubset: buttonType });
        if (buttonType === 'major')
            this.trigger('easyMajor', false);
        else if (buttonType === 'minor')
            this.trigger('easyMinor', false);
        else
            this.trigger('allChords', false);
    }

    isTriggered(buttonQual) {
        let check;
        if (this.props.api.lastHighScoreAccessed)
            check = this.props.api.lastHighScoreAccessed;
        else if (this.state.mode)
            check = this.state.mode + (this.state.inversions ? 'Inv' : '');
        else
            check = this.props.gameType + (this.props.inv ? 'Inv' : '');

        return (buttonQual === check) ? { backgroundColor: 'lime' } : {};
    }

    trigger(mode, inversions) {
        this.setState({ mode, inversions });
        this.props.dispatch(actions.getHighScores(mode + (inversions ? 'Inv' : '')));
    }

    componentDidMount(gT=this.props.gameType) {
        this.props.dispatch(actions.getHighScores(
            this.props.gameType + (this.props.inv ? 'Inv' : '')));

        if (/Minor/.test(gT))
            this.setState({ showButtonSubset: 'minor' });
        else if (/Major/.test(gT))
            this.setState({ showButtonSubset: 'major' });
        else
            this.setState({ showButtonSubset: 'all' });
    }

    render() {
        return (
            <div>

              Chord subsets
              <button onClick={() => this.clickBigButton('major')} style={this.bigButtonIsSelected('major')}>Major</button>
              <button onClick={() => this.clickBigButton('minor')} style={this.bigButtonIsSelected('minor')}>Minor</button>
              <button onClick={() => this.clickBigButton('all')} style={this.bigButtonIsSelected('all')}>All our chords</button>

              {this.state.showButtonSubset === 'major' ? (
                  <div>
                    <button onClick={() => this.trigger('easyMajor', false)} style={this.isTriggered('easyMajor')}>Easy</button>
                    <button onClick={() => this.trigger('easyMajor', true)} style={this.isTriggered('easyMajorInv')}>Easy, inversions</button>
                    <button onClick={() => this.trigger('hardMajor', false)} style={this.isTriggered('hardMajor')}>Hard</button>
                    <button onClick={() => this.trigger('hardMajor', true)} style={this.isTriggered('hardMajorInv')}>Hard, inversions</button>
                  </div>
              ) : null}

              {this.state.showButtonSubset === 'minor' ? (
                  <div>
                    <button onClick={() => this.trigger('easyMinor', false)} style={this.isTriggered('easyMinor')}>Easy</button>
                    <button onClick={() => this.trigger('easyMinor', true)}  style={this.isTriggered('easyMinorInv')}>Easy, inversions</button>
                    <button onClick={() => this.trigger('intermediateMinor', false)} style={this.isTriggered('intermediateMinor')}>Intermediate</button>
                    <button onClick={() => this.trigger('intermediateMinor', true)} style={this.isTriggered('intermediateMinorInv')}>Intermediate, inversions</button>
                    <button onClick={() => this.trigger('hardMinor', false)} style={this.isTriggered('hardMinor')}>Hard</button>
                    <button onClick={() => this.trigger('hardMinor', true)} style={this.isTriggered('hardMinorInv')}>Hard, inversions</button>
                  </div>
              ) : null}

              {this.state.showButtonSubset === 'all' ? (
                  <div>
                    <button onClick={() => this.trigger('allChords', false)} style={this.isTriggered('allChords')}>All our chords</button>
                    <button onClick={() => this.trigger('allChords', true)} style={this.isTriggered('allChordsInv')}>All our chord, inversions</button>
                  </div>
              ) : null}

              <ScoresTable tableType="highScore"/>

            </div>
        );
    }

}

const mapStateToProps = (state, props) => {
    let lastHighScore, inv;
    if (state.api.lastHighScoreAccessed) {
        lastHighScore = state.api.lastHighScoreAccessed.split('Inv');
        inv = lastHighScore.length === 2;
        lastHighScore = lastHighScore[0];
    }

    return {
        gameType: lastHighScore || state.game.gameType || 'allChords',
        inv: inv || state.game.inversions,
        api: state.api || { highScores: [] },
    };
};

export default connect(mapStateToProps)(HighScoresDisplay);
