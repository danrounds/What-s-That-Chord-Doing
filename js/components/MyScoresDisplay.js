import React from 'react';
import {connect} from 'react-redux';

import * as actions from '../actions';
import ScoresTable from './ScoresTable';

export class MyScoresDisplay extends React.Component {
    componentDidMount() {
        this.props.dispatch(actions.getMyScores(this.props.api.authToken));
    }

    render() {
        const name = this.props.api.name;
        let hText = name ?
                `Scores for ${this.props.api.name}` : 'Log in for user scores';

        return (
            <div>
              <div className="scores-btn-div">
                <h1 className="scores-h-txt">{hText}</h1>
              </div>
              <ScoresTable tableType="myScores" />
            </div>
        );
    }
}

export default connect((state) => ({ api: state.api }))(MyScoresDisplay);
