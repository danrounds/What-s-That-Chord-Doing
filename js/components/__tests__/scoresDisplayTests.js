import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { HighScoresDisplay } from '../HighScoresDisplay'; // gameType, inv, api.highScores, api.lastHighScoreAccessed;
import { MyScoresDisplay } from '../MyScoresDisplay'; // api.name = 'Somename'; api.authToken...?; 
import { ScoresTable } from '../ScoresTable'; // stub out some API data for display

// These are the tests for the subcomponents of of our Score component

const api = {
    name: 'davey',
    lastHighScoreAccessed: 'hardMajor',
    pending: false,
    error: false,
    myScores: {
        'name': 'davey',
        'scores': {
            'allChordsInv': {
                'totalClicks': 22,
                'nAnsweredRight': 3,
                'nQuestionNumber': 10,
                'winRatio': 0.3
            },
            'easyMajor': {
                'totalClicks': 12,
                'nAnsweredRight': 4,
                'nQuestionNumber': 10,
                'winRatio': 0.4
            }
        }
    },

    highScores: [
        {
            'name': 'ijijijij',
            'scores': {
                'intermediateMinor': {
                    'totalClicks': 47,
                    'nAnsweredRight': 10,
                    'nQuestionNumber': 10,
                    'winRatio': 1
                }
            }
        },
        {
            'name': 'bobby4',
            'scores': {
                'intermediateMinor': {
                    'totalClicks': 12346,
                    'nAnsweredRight': 124,
                    'nQuestionNumber': 129,
                    'winRatio': 0.9612403100775194
                }
            }
        },
        {
            'name': 'aseofj_',
            'scores': {
                'intermediateMinor': {
                    'totalClicks': 43,
                    'nAnsweredRight': 9,
                    'nQuestionNumber': 10,
                    'winRatio': 0.9
                }
            }
        }
    ]
};

describe('Score-based components', () => {
    // This only checks the "behavior" of part of our tree, but it's quick,
    // dirty, and gives us a reasonable idea that the component is working
    test('HighScoresDisplay', () => {
        const dispatch = (arg) => null;
        const comp = shallow(
            <HighScoresDisplay
               dispatch={dispatch}
               gameType="hardMajor"
               inv={true}
               api={api}
               />);
        comp.setState({ showButtonSubset: 'major' });
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('MyScoresDisplay', () => {
        const dispatch = (arg) => null;
        const comp = shallow(
            <MyScoresDisplay
               dispatch={dispatch}
               api={{ name: 'davey' }}
               />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('ScoresTable', () => {
        const comp = shallow(<ScoresTable tableType="myScores" api={api} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

});
