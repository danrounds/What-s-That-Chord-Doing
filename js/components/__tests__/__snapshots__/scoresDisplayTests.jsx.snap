// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score-based components HighScoresDisplay 1`] = `
<div>
  <div
    className="scores-btn-div"
  >
    <h1
      className="scores-h-txt"
    >
      Chord subsets
    </h1>
    <div
      className="big-btn-div"
    >
      <button
        className="big-btn big-btn-selected"
        onClick={[Function]}
      >
        Major
      </button>
      <button
        className="big-btn"
        onClick={[Function]}
      >
        Minor
      </button>
      <button
        className="big-btn"
        onClick={[Function]}
      >
        All our chords
      </button>
    </div>
    <div
      className="small-btns small-btns-major"
    >
      <button
        className="small-btn"
        onClick={[Function]}
      >
        Easy
      </button>
      <button
        className="small-btn"
        onClick={[Function]}
      >
        Easy
        <br />
        inversions
      </button>
      <button
        className="small-btn small-btn-selected"
        onClick={[Function]}
      >
        Hard
      </button>
      <button
        className="small-btn"
        onClick={[Function]}
      >
        Hard
        <br />
        inversions
      </button>
    </div>
  </div>
  <Connect(ScoresTable)
    tableType="highScore"
  />
</div>
`;

exports[`Score-based components MyScoresDisplay 1`] = `
<div>
  <div
    className="scores-btn-div"
  >
    <h1
      className="scores-h-txt"
    >
      Scores for davey
    </h1>
  </div>
  <Connect(ScoresTable)
    tableType="myScores"
  />
</div>
`;

exports[`Score-based components ScoresTable 1`] = `
<table
  className="scores-container"
>
  <thead>
    <tr>
      <th
        key="Game type"
      >
        Game type
      </th>
      <th
        key="?s Right"
      >
        ?s Right
      </th>
      <th
        key="?s Total"
      >
        ?s Total
      </th>
      <th
        key="Guesses per Right Answer"
      >
        Guesses per Right Answer
      </th>
      <th
        key="% Right"
      >
        % Right
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="0"
    >
      <th
        key="1"
      >
        Easy major
      </th>
      <th
        key="2"
      >
        4
      </th>
      <th
        key="3"
      >
        10
      </th>
      <th
        key="4"
      >
        3.33
      </th>
      <th
        key="5"
      >
        40
      </th>
    </tr>
    <tr
      key="1"
    >
      <th
        key="2"
      >
        All our chords, w/inversions
      </th>
      <th
        key="3"
      >
        3
      </th>
      <th
        key="4"
      >
        10
      </th>
      <th
        key="5"
      >
        1.36
      </th>
      <th
        key="6"
      >
        30
      </th>
    </tr>
  </tbody>
</table>
`;
